### PR TITLE
[MediaBundle] Restructured cleanup mechanism for Files

### PR DESCRIPTION
**Pull request by FabianLiebl**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Doc PR?       | no
| Backport      | 0.11
| License       | MIT

Restructured cleanup mechanism for Files

### CODE REVIEW

File: src/Enhavo/Bundle/AppBundle/Util/StopWatch.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace Enhavo\Bundle\AppBundle\Util;
+
+class StopWatch
```
Comment (gseidel) [inline src/Enhavo/Bundle/AppBundle/Util/StopWatch.php:5]: Please use symfony stop watch instead.

https://symfony.com/doc/5.0/components/stopwatch.html

File: src/Enhavo/Bundle/DoctrineExtensionBundle/Util/AssociationFinder.php [hunk 2]
```diff
@@ -7,21 +7,19 @@
...
      */
-    public function __construct(EntityManagerInterface $em)
-    {
-        $this->em = $em;
-    }
+    public function __construct(
+        private EntityManagerInterface $entityManager
+    ) {}
+
+    #region Associations to
```
Comment (gseidel) [inline src/Enhavo/Bundle/DoctrineExtensionBundle/Util/AssociationFinder.php:22]: Please don't use regions in enhavo. Remove all occurrences

File: src/Enhavo/Bundle/MediaBundle/GarbageCollection/GarbageCollector.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+            if ($result === GarbageCollectionVoterInterface::VOTE_KEEP) {
+                return false;
+            }
+        }
+
+        return false;
+    }
+
+    protected function deleteFile(FileInterface $file) {
```
Comment (gseidel) [inline src/Enhavo/Bundle/MediaBundle/GarbageCollection/GarbageCollector.php:127]: Code style. Curly brackets in next line for functions

File: src/Enhavo/Bundle/MediaBundle/GarbageCollection/GarbageCollector.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        $this->doRun($maxItems, $andFlush);
+    }
+
+    public function dryRun(?int $maxItems = null, bool $andFlush = true): void
+    {
+        $this->dryRun = true;
+        $this->doRun($maxItems, $andFlush);
+    }
+
+    protected function doRun(?int $maxItems = null, bool $andFlush = true): void
```
Comment (gseidel) [inline src/Enhavo/Bundle/MediaBundle/GarbageCollection/GarbageCollector.php:45]: Please use private functions in this class

File: src/Enhavo/Bundle/MediaBundle/GarbageCollection/Voter/ReferenceGarbageCollectionVoter.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    ) {}
+
+    public function vote(FileInterface $file): string
+    {
+        if (!$this->enableDeleteUnreferenced) {
+            return self::VOTE_ABSTAIN;
+        }
+
+        // Prevent deleting files that have just been uploaded and just aren't connected yet
+        if ($this->existenceTimestampThreshold === null) {
```
Comment (gseidel) [inline src/Enhavo/Bundle/MediaBundle/GarbageCollection/Voter/ReferenceGarbageCollectionVoter.php:25]: Do we need that? For this case we have the garbage flag.

**Review by gseidel — CHANGES_REQUESTED** [review]

File: src/Enhavo/Bundle/DoctrineExtensionBundle/Util/AssociationFinder.php [hunk 6]
```diff
@@ -134,13 +118,99 @@ private function buildAssociationMapCache($targetClass, $excludedClasses = [])
...
+            $entityClassName = get_class($entity);
+        }
+        if (!($entity instanceof $entityClassName)) {
+            throw new \Exception('Error: Parameter $entityClassName must be the fully qualified class name of $entity, one of its parent classes or interfaces, or null.');
+        }
+
+        $results = [];
+        foreach($this->getOutgoingAssociationMap($entityClassName, $excludeClasses) as $association) {
+
+            //TODO?
```
[the remark this answers is not in the document]
Comment (FabianLiebl) [reply src/Enhavo/Bundle/DoctrineExtensionBundle/Util/AssociationFinder.php]: Sorry, I forgot this. This method is actually unfinished, but it's not needed anyhow. Removed it.

**Review by gseidel — APPROVED** [review]

**Pull request merged**